src/Forms/ResourceLocatorField.php

```diff
@@ -0,0 +1,272 @@
+<?php
+namespace SilverStripe\CKANRegistry\Forms;
+
+use InvalidArgumentException;
+use SilverStripe\Forms\FormField;
+use SilverStripe\ORM\DataObjectInterface;
+
+class ResourceLocatorField extends FormField
+{
+    /**
+     * The default CKAN endpoint to be used in a default isn't provided on construction
+     * @see ResourceLocatorField::$defaultEndpoint
+     *
+     * @config
+     * @var string
+     */
+    private static $default_endpoint = 'https://catalogue.data.govt.nz/';
+
+    /**
+     * The default CKAN endpoint to be used in this field. This will allow consumers of the field to only provide
+     * package or dataset IDs and still work. If not set the configured default will instead be used.
+     *
+     * @var string|null
+     */
+    protected $defaultEndpoint;
+
+    /**
+     * Set a site name that can be used to refer to the CKAN endpoint. By default this will be "a CKAN website".
+     *
+     * @var string|null
+     */
+    protected $siteName = null;
+
+    /**
+     * The name of the subfield to save the endpoint value of this field into
+     *
+     * @var string
+     */
+    protected $endpointFieldName = 'Endpoint';
+
+    /**
+     * The name of the subfield to save the dataset value of this field into
+     *
+     * @var string
+     */
+    protected $datasetFieldName = 'DataSet';
+
+    /**
+     * The name of the subfield to save the resource value of this field into
+     *
+     * @var string
+     */
+    protected $resourceFieldName = 'Resource';
+
+    /**
+     * @param string $name
+     * @param string $title
+     * @param string $value
+     * @param string $defaultEndpoint
+     */
+    public function __construct($name, $title = null, $value = null, $defaultEndpoint = null)
+    {
+        parent::__construct($name, $title, $value);
+        $this->setDefaultEndpoint($defaultEndpoint);
+
+        // Set a default description
+        $this->setDescription(_t(
+            __CLASS__ . '.DESCRIPTION',
+            'Connect to a data source from {site}. Once added and saved you can configure the appearance and add search'
+            . ' filters.',
+            [ 'site' => $this->getSiteName() ]
+        ));
+
+        $this->addExtraClass('ckan-resource-locator__container');
+    }
+
+    public function getSchemaDataDefaults()
+    {
+        $schemaData = parent::getSchemaDataDefaults();
+
+        $schemaData['defaultEndpoint'] = $this->getDefaultEndpoint();
+
+        return $schemaData;
+    }
+
+    public function setValue($value, $data = null)
+    {
+        // $value should be the child DataObject but if it isn't we can use `getSaveTarget`
+        if (!$value instanceof DataObjectInterface) {
+            $value = $this->getSaveTarget($data);
```

Comment:
There's a bug here - $data is nullable AND it unlikely to be a DataObjectInterface, but getSaveTarget has that interface type hinted

Reply:
Well spotted. Yeah we should assert that `$data instanceof DataObjectInterface` too. I'll follow up as part of the work I do with #18.